### DISCUSSION TITLE
CSS and Javascript MIME Errors

### DISCUSSION THREAD

**Discussion by Junto026** [authelia/authelia#6106] [Troubleshooting] [answered]
I'm having an issue accessing another app (Plex) through NGINX when Authelia is being used to protect Plex.  If I remove Authelia, the issue goes away.  I don't understand the error, and I'm sure there's a configuration change I can make to resolve it, but am not sure where to start.

Screenshot of the error.  It happens after I authenticate with Authelia and it then tries to re-direct me to Plex.
<img width="1792" alt="Screenshot 2023-10-08 at 8 36 35 PM" src="https://github.com/authelia/authelia/assets/88695895/de20088a-5167-44a8-81db-6602e2e4703d">

Often, if I refresh a few times, the error goes away and Plex loads.

Should I remove the `X-Content-Type-Options: no sniff` or is that a terrible idea?  If I should remove it, where?  (I can't find where that is being added).  Other possible causes / suggestions I should research?

**configuration.yaml**

```
###############################################################################
#                           Authelia Configuration                            #
###############################################################################

## Note: the container by default expects to find this file at authelia:/config/configuration.yml

## Certificates directory specifies where Authelia will load trusted certificates (public portion) from in addition to
## the system certificates store.
## They should be in base64 format, and have one of the following extensions: *.cer, *.crt, *.pem.
# certificates_directory: /config/certificates/

theme: dark
jwt_secret: <redacted>
default_redirection_url: https://authelia.mydomain.com
default_2fa_method: "webauthn" ## Options are totp, webauthn, mobile_push.

server:
  host: 0.0.0.0
  port: 9091
  path: ""
  # asset_path: /config/assets/
  enable_pprof: false
  enable_expvars: false
  disable_healthcheck: false

  tls:
  ## Authelia by default doesn't accept TLS communication on the server port. This section overrides this behaviour.
    key: "/config/certificates/privkey.pem" ## The path to the DER base64/PEM format private key.
    certificate: "/config/certificates/fullchain.pem" ## The path to the DER base64/PEM format public certificate.
    client_certificates: [] ## The list of certificates for client authentication.

  headers:
  ## Server headers configuration/customization.
    csp_template: "" ## The CSP Template. Read the docs.

  # buffers:
  ## Server Buffers configuration.
    # read: 4096 ## Read buffer.
    # write: 4096 ## Write buffer.

  # timeouts:
  ## Server Timeouts configuration.
    # read: 6s ## Read timeout.
    # write: 6s ## Write timeout.
    # idle: 30s ## Idle timeout.

log:
  level: trace ## Level of verbosity for logs: error, warn, info, debug, trace.
#  format: json ## Format the logs are written as: json, text.
#  file_path: /config/authelia.log ## File path where the logs will be written. If not set logs are written to stdout.
#  keep_stdout: true ## Whether to also log to stdout when a log_file_path is defined.

totp:
  disable: true ## Disable TOTP.
  issuer: my domain.com ## The issuer name displayed in the Authenticator application of your choice.
  algorithm: sha1 ## The TOTP algorithm to use.
  digits: 6 ## The number of digits a user has to input. Must either be 6 or 8.
  period: 300   ## The period in seconds a one-time password is valid for.
  skew: 1 ## The skew controls number of one-time passwords either side of the current one that are valid.
  secret_size: 32 ## The size of the generated shared secrets. Default is 32 and is sufficient in most use cases, minimum is 20.

webauthn:
  disable: false  
  timeout: 60s ## Adjust the interaction timeout for Webauthn dialogues.
  display_name: Authelia ## The display name the browser should show the user for when using Webauthn to login/register.  
  attestation_conveyance_preference: none  ## Options are none, indirect, direct.  Controls if attestation statement including the AAGUID is collected
  user_verification: preferred  ## Options are required, preferred, discouraged.  Controls if hte user must make a gesture or action to confirm they are present.

duo_api:
  disable: true
  # hostname: api-123456789.example.com
  # integration_key: ABCDEF
  ## Secret can also be set using a secret: https://www.authelia.com/c/secrets
  # secret_key: 1234567890abcdefghifjkl
  # enable_self_enrollment: false

ntp:
  address: "time.cloudflare.com:123" ## NTP server address.
  version: 4 ## NTP version.
  max_desync: 3s ## Maximum allowed time offset between the host and the NTP server.
  disable_startup_check: false ## Disables the NTP check on startup entirely. This means Authelia will not contact a remote service at all if you
  disable_failure: false   ## The default of false will prevent startup only if we can contact the NTP server and the time is out of sync

authentication_backend:
## The available providers are: `file`, `ldap`. You must use only one of these providers.
  password_reset:
    disable: false ## Disable both the HTML element and the API for reset password functionality.
    custom_url: "" ## External reset password url that redirects the user to an external reset portal. This disables the internal reset functionality.

  refresh_interval: 5m ## The amount of time to wait before we refresh data from the authentication backend. Uses duration notation.

#  file: 
#  ## File (Authentication Provider)
#    path: /config/users_database.yml
#    watch: false
#    search:
#      email: false
#      case_insensitive: false
#    password:
#      algorithm: argon2id
#      iterations: 3
#      memory: 64
#      parallelism: 4
#      key_length: 32
#      salt_length: 16

  ldap:
    implementation: custom
    url: ldaps://openldap.mydomain.com
#    url: ldap://192.168.32.105
    timeout: 5s ## The dial timeout for LDAP.
    start_tls: false ## Use StartTLS with the LDAP connection.
    tls:
      server_name: openldap.mydomain.com
      skip_verify: true ## Skip verifying the server certificate entirely. 
      minimum_version: TLS1.2
      maximum_version: TLS1.3

    base_dn: dc=mydomain,dc=com
    username_attribute: uid
    additional_users_dn: ou=users
    users_filter: (&({username_attribute}={input})(objectClass=inetOrgPerson))
    additional_groups_dn: ou=groups
    groups_filter: (&(member={dn})(objectClass=groupOfNames))
    group_name_attribute: cn
    mail_attribute: mail
    display_name_attribute: displayName
    permit_referrals: false ## Follow referrals returned by the server. This is especially useful for environments where read-only servers exist. Only implemented for write operations.
    user: cn=admin,dc=mydomain,dc=com
    password: <redacted> 

password_policy:
  standard:
    enabled: true
    min_length: 12 
    max_length: 0  
    require_uppercase: true 
    require_lowercase: true 
    require_number: true 
    require_special: true

access_control:
  default_policy: deny ## Default policy can either be 'bypass', 'one_factor', 'two_factor' or 'deny'. 
  rules:
    - domain: 'authelia.mydomain.com'
      policy: bypass
    - domain: '*.mydomain.com'
      policy: two_factor
      subject:
      - 'group:autheliaadmins'

session:
  name: authelia_session ## The name of the session cookie.
  domain: my domain.com
  same_site: lax ## Sets the Cookie SameSite value. Possible options are none, lax, or strict.
  secret: insecure_session_secret ## The secret to encrypt the session data. This is only used with Redis / Redis Sentinel.
  expiration: 1M ## The time before the cookie expires and the session is destroyed if remember me IS NOT selected.
  inactivity: 1M
  remember_me_duration: -1 ## 1M default; Value of -1 disables remember me.

regulation:
  max_retries: 3 ## The number of failed login attempts before user is banned. Set it to 0 to disable regulation.
  find_time: 2m ## The time range during which the user can attempt login before being banned. The user is banned if the authentication failed 'max_retries' times in a 'find_time' seconds window. Find Time accepts duration notation.
  ban_time: 5m ## The length of time before a banned user can login again. Ban Time accepts duration notation.

storage:
## The available providers are: `local`, `mysql`, `postgres`. You must use one and only one of these providers.
  encryption_key: <redacted>
  local:
  ## Path to the SQLite3 Database.
    path: /config/db.sqlite3

notifier:
  disable_startup_check: true  ## You can disable the notifier startup check by setting this to true.

  smtp:
    host: smtp.mail.me.com
    port: 587
    timeout: 5s
    username: user@mydomain.com
    password: <redacted>
    sender: "Authelia <services@mydomain.com>"
    # identifier: localhost. ## HELO/EHLO Identifier. Some SMTP Servers may reject the default of localhost.
    subject: "{title}"
    startup_check_address: user@mydomain.com   
    # disable_require_tls: false  ## By default we require some form of TLS. This disables this check though is not advised.
    # disable_html_emails: false

    tls:
      skip_verify: false
      minimum_version: TLS1.2
...
```

**nginx.conf Snippet**
```
	# Authelia
	server {
		listen 443 ssl;
		listen [::]:443 ssl;
		http2 on;
		server_name authelia.mydomain.com;
		include /etc/nginx/conf.d/certificate_locations;

		location / {
			proxy_pass https://192.168.32.102:9091;
			include /etc/nginx/conf.d/authelia_nginx_config;
		}
	}

	# Plex Server
	server {
		listen 443 ssl;
		listen [::]:443 ssl;
		http2 on;
		server_name plex.mydomain.com;
		include /etc/nginx/conf.d/authelia_protectedapp_connection;

		location / {
			location = / {
 				return 301 /web;
			}			

			proxy_pass https://192.168.32.106:32400; 
			include /etc/nginx/conf.d/authelia_protectedapp_config;	
		}
	}
```

**/etc/nginx/conf.d/authelia_nginx_config**
```

## Headers
proxy_set_header Host $host;
proxy_set_header X-Original-URL $scheme://$http_host$request_uri;
proxy_set_header X-Forwarded-Proto $scheme;
proxy_set_header X-Forwarded-Host $http_host;
proxy_set_header X-Forwarded-Uri $request_uri;
proxy_set_header X-Forwarded-Ssl on;
proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
proxy_set_header X-Real-IP $remote_addr;
proxy_set_header Connection "";

## Basic Proxy Configuration
client_body_buffer_size 128k;
proxy_next_upstream error timeout invalid_header http_500 http_502 http_503; ## Timeout if the real server is dead.
proxy_redirect  http://  $scheme://;
proxy_http_version 1.1;
proxy_cache_bypass $cookie_session;
proxy_no_cache $cookie_session;
proxy_buffers 64 256k;

## Trusted Proxies Configuration
## Please read the following documentation before configuring this:
##     https://www.authelia.com/integration/proxies/nginx/#trusted-proxies
set_real_ip_from 192.168.32.0/20;
set_real_ip_from 10.0.0.0/24;
real_ip_header X-Forwarded-For;
real_ip_recursive on;

## Advanced Proxy Configuration
send_timeout 5m;
proxy_read_timeout 360;
proxy_send_timeout 360;
proxy_connect_timeout 360;

```

**/etc/nginx/conf.d/authelia_protectedapp_connection**
```
location /authelia {
	proxy_pass https://192.168.32.102:9091/api/verify;    
	internal;
	proxy_pass_request_body off;
	proxy_set_header Content-Length "";

	# Timeout if the real server is dead
	proxy_next_upstream error timeout invalid_header http_500 http_502 http_503;
	client_body_buffer_size 128k;
	proxy_set_header Host $host;
	proxy_set_header X-Original-URL $scheme://$http_host$request_uri;
	proxy_set_header X-Real-IP $remote_addr;
	proxy_set_header X-Forwarded-For $remote_addr; 
	proxy_set_header X-Forwarded-Proto $scheme;
	proxy_set_header X-Forwarded-Host $http_host;
	proxy_set_header X-Forwarded-Uri $request_uri;
	proxy_set_header X-Forwarded-Ssl on;
	proxy_redirect  http://  $scheme://;
	proxy_http_version 1.1;
	proxy_set_header Connection "";
	proxy_cache_bypass $cookie_session;
	proxy_no_cache $cookie_session;
	proxy_buffers 4 32k;

	send_timeout 5m;
	proxy_read_timeout 240;
	proxy_send_timeout 240;
	proxy_connect_timeout 240;
}
```

**/etc/nginx/conf.d/authelia_protectedapp_config**
```

auth_request /authelia;
auth_request_set $target_url https://$http_host$request_uri;
auth_request_set $user $upstream_http_remote_user;
auth_request_set $email $upstream_http_remote_email;
auth_request_set $groups $upstream_http_remote_groups;
proxy_set_header Remote-User $user;
proxy_set_header Remote-Email $email;
proxy_set_header Remote-Groups $groups;

error_page 401 =302 https://authelia.mydomain.com/?rd=$target_url; #change this to match your authentication domain/subdomain

client_body_buffer_size 128k;

proxy_next_upstream error timeout invalid_header http_500 http_502 http_503;

send_timeout 5m;
proxy_read_timeout 360;
proxy_send_timeout 360;
proxy_connect_timeout 360;

proxy_set_header Host $host;
proxy_set_header Upgrade $http_upgrade;
proxy_set_header Connection "Upgrade";
proxy_set_header Accept-Encoding gzip;
proxy_set_header X-Real-IP $remote_addr;
proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
proxy_set_header X-Forwarded-Proto $scheme;
proxy_set_header X-Forwarded-Host $http_host;
proxy_set_header X-Forwarded-Uri $request_uri;
proxy_set_header X-Forwarded-Ssl on;
proxy_redirect  http://  $scheme://;
proxy_http_version 1.1;
proxy_set_header Connection "";
proxy_cache_bypass $cookie_session;
proxy_no_cache $cookie_session;
proxy_buffers 64 256k;

set_real_ip_from 192.168.32.0/20;
set_real_ip_from 10.0.0.0/24;
real_ip_header X-Forwarded-For;
real_ip_recursive on;
```

**Comment by Junto026** [comment]
I found the below in the Authelia logs when the issue occurs...

> 2023-10-09 14:59:40 authelia  | time="2023-10-09T14:59:40-04:00" level=info msg="Access to https://plex.mydomain.com/web/js/3519-3519-35b03a3b034aaa63bb09-plex-4.108.0.32184-ffafb0d.js (method unknown) is not authorized to user <anonymous>, responding with status code 401" method=GET path=/api/verify remote_ip=192.168.65.1
> 2023-10-09 14:59:40 authelia  | time="2023-10-09T14:59:40-04:00" level=trace msg="Checking if we need check the authentication backend for an updated profile for ." method=GET path=/api/verify remote_ip=192.168.65.1
> 2023-10-09 14:59:40 authelia  | time="2023-10-09T14:59:40-04:00" level=debug msg="Check authorization of subject username= groups= ip=192.168.65.1 and object https://plex.mydomain.com/web/main-179-69d37121295f158388fa-plex-4.108.0.32184-ffafb0d.css (method )."
> 2023-10-09 14:59:40 authelia  | time="2023-10-09T14:59:40-04:00" level=trace msg="Replied (status=401)" method=GET path=/api/verify remote_ip=192.168.65.1
> 2023-10-09 14:59:40 authelia  | time="2023-10-09T14:59:40-04:00" level=trace msg="ACL MISS Position 1 for subject username= groups= ip=192.168.65.1 and object https://plex.mydomain.com/web/main-179-69d37121295f158388fa-plex-4.108.0.32184-ffafb0d.css (method )"
> 2023-10-09 14:59:40 authelia  | time="2023-10-09T14:59:40-04:00" level=trace msg="ACL HIT Position 2 for subject username= groups= ip=192.168.65.1 and object https://plex.mydomain.com/web/main-179-69d37121295f158388fa-plex-4.108.0.32184-ffafb0d.css (method )"
> 2023-10-09 14:59:40 authelia  | time="2023-10-09T14:59:40-04:00" level=trace msg="ACL MISS Position 1 for subject username= groups= ip=192.168.65.1 and object https://plex.mydomain.com/web/js/main-179-69d37121295f158388fa-plex-4.108.0.32184-ffafb0d.js (method )"
> 2023-10-09 14:59:40 authelia  | time="2023-10-09T14:59:40-04:00" level=trace msg="ACL HIT Position 2 for subject username= groups= ip=192.168.65.1 and object https://plex.mydomain.com/web/js/main-179-69d37121295f158388fa-plex-4.108.0.32184-ffafb0d.js (method )"
> 2023-10-09 14:59:40 authelia  | time="2023-10-09T14:59:40-04:00" level=info msg="Access to https://plex.mydomain.com/web/js/main-179-69d37121295f158388fa-plex-4.108.0.32184-ffafb0d.js (method unknown) is not authorized to user <anonymous>, responding with status code 401" method=GET path=/api/verify remote_ip=192.168.65.1
> 2023-10-09 14:59:40 authelia  | time="2023-10-09T14:59:40-04:00" level=trace msg="Replied (status=401)" method=GET path=/api/verify remote_ip=192.168.65.1
> 2023-10-09 14:59:40 authelia  | time="2023-10-09T14:59:40-04:00" level=info msg="Access to https://plex.mydomain.com/web/main-179-69d37121295f158388fa-plex-4.108.0.32184-ffafb0d.css (method unknown) is not authorized to user <anonymous>, responding with status code 401" method=GET path=/api/verify remote_ip=192.168.65.1
> 2023-10-09 14:59:40 authelia  | time="2023-10-09T14:59:40-04:00" level=trace msg="Replied (status=401)" method=GET path=/api/verify remote_ip=192.168.65.1
> 2023-10-09 14:59:40 authelia  | time="2023-10-09T14:59:40-04:00" level=trace msg="ACL HIT Position 2 for subject username= groups= ip=192.168.65.1 and object https://plex.mydomain.com/web/3519-3519-35b03a3b034aaa63bb09-plex-4.108.0.32184-ffafb0d.css (method )"
> 2023-10-09 14:59:40 authelia  | time="2023-10-09T14:59:40-04:00" level=info msg="Access to https://plex.mydomain.com/web/3519-3519-35b03a3b034aaa63bb09-plex-4.108.0.32184-ffafb0d.css (method unknown) is not authorized to user <anonymous>, responding with status code 401" method=GET path=/api/verify remote_ip=192.168.65.1
> 2023-10-09 14:59:40 authelia  | time="2023-10-09T14:59:40-04:00" level=trace msg="Replied (status=401)" method=GET path=/api/verify remote_ip=192.168.65.1
> 2023-10-09 14:59:41 authelia  | time="2023-10-09T14:59:41-04:00" level=trace msg="Request hit" method=GET path=/api/health remote_ip=127.0.0.1
> 2023-10-09 14:59:41 authelia  | time="2023-10-09T14:59:41-04:00" level=trace msg="Replied (status=200)" method=GET path=/api/health remote_ip=127.0.0.1

**Reply by james-d-elliott** [reply, marked as answer]
Looks like either the cookie is not being passed by plex when requesting those assets or the plex js is not following redirects (or both) which leads to the issue.You can try bypassing the js directory but as to what other issues that might result in I'm not sure.

**Reply by Junto026** [reply]
Interestingly, bypassing the plex.mydomain.com/web/js directory did not fix it.  But bypassing the entire plex.mydomain.com/web/ directory did.

I don't love the "fix", but the Plex app does have its own built-in authentication and 2FA, so I suppose it's OK.

The issue is funky since it only happens on Safari.  And disabling all cookie/privacy features in Safari and MacOS didn't help.

**Reply by james-d-elliott** [reply]
Very unusual. Can you verify in the requests in the browser that it's not sending the cookie header? I'm not sure we can do much if it's excluding it except ask them to include it. I suspect when it occurs the cookie is absent and when it doesn't it's present. Also does this only occur with Safari + Plex? If so maybe it's a Safari bug.

**Reply by Junto026** [reply]
This is what I see in network transactions when the issue occurs.  It looks to me no Cookies are being provided by Safari.  Does this indicate the same to you?  Yes I'm only seeing it with Safari+Plex... possibly it is a Safari bug.

![Screenshot 2023-10-10 at 9 17 44 AM](https://github.com/authelia/authelia/assets/88695895/49464bd1-6bdb-4f4a-b3f6-9562cf72385b)
![Screenshot 2023-10-10 at 9 17 42 AM](https://github.com/authelia/authelia/assets/88695895/742ceaea-92c3-4ead-92cc-41eaa18263dd)
![Screenshot 2023-10-10 at 9 17 39 AM](https://github.com/authelia/authelia/assets/88695895/d22197e3-658a-4bb0-9f6d-3ef7fdf5aaf3)
![Screenshot 2023-10-10 at 9 17 37 AM](https://github.com/authelia/authelia/assets/88695895/b21e919c-b8ee-467a-ae38-815343c568f5)
![Screenshot 2023-10-10 at 9 17 35 AM](https://github.com/authelia/authelia/assets/88695895/825169e1-30cf-490e-99b3-a07f8b07f4e8)
![Screenshot 2023-10-10 at 9 17 33 AM](https://github.com/authelia/authelia/assets/88695895/063925c3-62be-433a-99cb-2db69a8a966f)
![Screenshot 2023-10-10 at 9 17 30 AM](https://github.com/authelia/authelia/assets/88695895/629ad5b2-a0b5-4510-8fb8-d4bebbc0addf)

**Reply by james-d-elliott** [reply]
Looking like it is indeed a bug with how Safari treats cookies: https://stackoverflow.com/questions/66324724/safari-set-cookies-not-working-for-first-party-cookie

Maybe try changing the [same_site](https://www.authelia.com/configuration/session/introduction/#same_site) setting.

**Reply by Junto026** [reply]
Seems like it could be it.  I set `same_site` to `none` and it seems to have resolved it :). We'll see if it re-appears after more use.

**Reply by james-d-elliott** [reply]
Unfortunate they don't follow the standard / specifications on such a critical aspect of cookies. Keep in mind using this value would theoretically make the cookie more likely to be divulged via a cross-site request/attack.

**Reply by AsteroidOrangeJuice** [reply]
I've just come across the same issue and would prefer not to set same_site to none - [apparently](https://stackoverflow.com/a/66448703) removing `max age` from the cookie can also fix this problem, but the configuration for Duration doesn't seem to support this. Is this possible at all?

**Reply by james-d-elliott** [reply]
Not setting a cookie expiration is not really an answer either I don't think if that's what you mean (we do not set the Max-Age value with the Set-Cookie header we set an example cookie value of `*******; expires=Mon, 08 Apr 2024 07:46:49 GMT; domain=example.com; path=/; HttpOnly; secure; SameSite=Lax`) due the the fact if you then close the browser the cookie is immediately deleted. The options are to exclude the assets themselves, or set the same site value.

**Reply by AsteroidOrangeJuice** [reply]
Unfortunately setting `same_site` to `none` also didn't fix this for me.

**Reply by james-d-elliott** [reply]
I'm not sure what to tell you other than to add bypass rules to the UI assets. We can't do anything additional as far as I'm aware to control if they include the cookie or not.

**Reply by AsteroidOrangeJuice** [reply]
css and certain images I wouldn't mind at all, but some of my javascript files I definitely wouldn't want exposed to the internet.

After more digging, I found that some of my problems were coming from trying to load a js file, it redirecting to Authelia, but then the response was just an html page saying I needed to enable javascript to actually get to Authelia. (I guess the js file shouldn't technically have gotten to the point of even being requested, as the initial webpage should have redirected instead and then the cookie should have been included in the js file request, but I think that was caused by the web application caching somehow. Some of the safari cookie handling weirdness likely comes into play here too.).

Anyways, point was: my javascript file was cached, but the contents of that cache were an html file, which gave me the same MIME errors called out here. I think this could be ~fixed~ mitigated one of two ways. I'm actually using both for testing right now:

Mitigations

1: In my authelia config in Nginx, I just disabled caching for everything coming from the Authelia server:

```
add_header Cache-Control 'no-store, no-cache' always;
location / {
   ... authelia config here
}
```

2: If I see a redirect request for js or css come to authelia, know it won't be processed in the way I want (because of the above scenario) and throw a 500: - edit, okay, don't use this one, it breaks things.

```
location / {
       set $shouldReject 1;
        if ($request_uri !~ "^\/\?rd=") {
            set $shouldReject 0;
        }
        if ($request_uri !~ "(\.css|\.js)$") {
            set $shouldReject 0;
        }
        if ($shouldReject) {
            return 500;
        }
        ... rest of authelia config here
}
```

**Reply by james-d-elliott** [reply]
Very odd. Appears like when being redirected plex is instead loading the content of the redirect site in place of the js/css file entirely. Also for reference there should be no risk in bypassing the JS as all plex installs should have the same JS bundles.

Either way it is fairly clear this is beyond our control, plex needs to configure their use of the fetchAPI to follow redirects properly. The actual fix will have to occur on their end, the workarounds are however appreciated.

**Reply by AsteroidOrangeJuice** [reply]
To be clear, I'm not the original poster - I'm encountering the same problem with a flask server as well as a https://rust-lang.github.io/mdBook/ application. 

> Appears like when being redirected plex is instead loading the content of the redirect site in place of the js/css file entirely.

Unless I'm misunderstanding something, I don't actually know how this should work - the browser requests a javascript file which is used on the html webpage - but because it's requested without the proper cookie, the request is sent to authelia, but because it's an asset used by the webpage, not actually the page the user is navigating too, javascript isn't loaded, and there's no actual redirect in the response.

I'm able to reproduce this in postman too - just request any css file without passing the appropriate auth, and you'll get a 200 with the HTML saying you need to enable javascript to run authelia.

**Reply by james-d-elliott** [reply]
Because it's sent without a cookie, authelia responds with a redirect, or in the case of nginx, nginx responds with the redirect. The application in question instead of following the redirect is just loading the authelia page in place of the originally requested file. What you're seeing in postman is the final redirect, if you disable following redirects or use curl then you will see what I mean.

**Reply by james-d-elliott** [reply]
See here for an example which very clearly shows there was a `location` response header with the proper redirect and status code:

```
curl -v https://whoami.example.com
* Host whoami.example.com:443 was resolved.
* IPv6: (none)
* IPv4: 139.99.145.66
*   Trying 139.99.145.66:443...
* Connected to whoami.example.com (139.99.145.66) port 443
* ALPN: curl offers h2,http/1.1
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: none
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_128_GCM_SHA256 / x25519 / RSASSA-PSS
* ALPN: server accepted h2
* Server certificate:
*  subject: CN=example.com
*  start date: Feb 15 03:33:36 2024 GMT
*  expire date: May 15 03:33:35 2024 GMT
*  subjectAltName: host "whoami.example.com" matched cert's "*.example.com"
*  issuer: C=US; O=Let's Encrypt; CN=R3
*  SSL certificate verify ok.
*   Certificate level 0: Public key type RSA (4096/152 Bits/secBits), signed using sha256WithRSAEncryption
*   Certificate level 1: Public key type RSA (2048/112 Bits/secBits), signed using sha256WithRSAEncryption
*   Certificate level 2: Public key type RSA (4096/152 Bits/secBits), signed using sha256WithRSAEncryption
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* using HTTP/2
* [HTTP/2] [1] OPENED stream for https://whoami.example.com/
* [HTTP/2] [1] [:method: GET]
* [HTTP/2] [1] [:scheme: https]
* [HTTP/2] [1] [:authority: whoami.example.com]
* [HTTP/2] [1] [:path: /]
* [HTTP/2] [1] [user-agent: curl/8.7.1]
* [HTTP/2] [1] [accept: */*]
> GET / HTTP/2
> Host: whoami.example.com
> User-Agent: curl/8.7.1
> Accept: */*
>
* Request completely sent off
< HTTP/2 302
< content-type: text/html; charset=utf-8
< date: Wed, 10 Apr 2024 07:40:47 GMT
< location: https://auth.example.com/authelia/?rd=https%3A%2F%2Fwhoami.example.com%2F&rm=GET
< permissions-policy: accelerometer=(), autoplay=(), camera=(), display-capture=(), geolocation=(), gyroscope=(), keyboard-map=(), magnetometer=(), microphone=(), midi=(), payment=(), picture-in-picture=(), screen-wake-lock=(), sync-xhr=(), xr-spatial-tracking=(), interest-cohort=()
< referrer-policy: strict-origin-when-cross-origin
< set-cookie: authelia-session=rGHflMs_x5B$J%hVjT51sGIF6fk%eGs3; expires=Wed, 10 Apr 2024 07:41:47 GMT; domain=example.com; path=/; HttpOnly; secure; SameSite=Lax
< x-content-type-options: nosniff
< x-dns-prefetch-control: off
< x-frame-options: DENY
< content-length: 118
<
* Connection #0 to host whoami.example.com left intact
```

**Reply by AsteroidOrangeJuice** [reply]
Right, but now curl `https://auth.example.com/authelia/?rd=https%3A%2F%2Fwhoami.example.com%2F&rm=GET` - that's where the browser is getting to. It follow the redirect to authelia, but because it's not an active browser window with javascript running, authelia doesn't run, and that html content including `You need to enable JavaScript to run this app.` is what is stored in place of the css/js/whatever asset.

**Reply by james-d-elliott** [reply]
If it was following the redirect it wouldn't load it instead of the file, it would actually follow the redirect within the viewport. This is what I'm saying. This is all expected if an application is not behaving per internet standards. Just like not including the cookie. It's a bug with the application.

Using fetch API you have to explicitly instruct it to follow redirects and include cookies I believe, for everything else this the default behavior. The fetch API is incredibly jank though in my opinion, wouldn't surprise me if this is their way of following redirects.

**Reply by james-d-elliott** [reply]
Actually looks like it's a bug/limitation with the proverbial junk that is fetch:

- https://github.com/whatwg/fetch/issues/763
- https://github.com/whatwg/fetch/issues/601

Combine that with the fact fetch doesn't include any distinguishable headers to determine the fact it is a fetch request unlike AJAX (which includes `Requested-With: XMLHttpRequest` which we handle using the known workaround i.e. sending a non 30x response, which is not overly compatible with nginx for other reasons), we're hamstrung here.

**Reply by AsteroidOrangeJuice** [reply]
So confirming my understanding:

If I load https://whoami.example.com without proper cookies, that redirects within the viewport to authelia, auth's fine, and redirects back just fine. All good here.

If https://whoami.example.com loads fine (for whatever reason, local caching or safari being goofy), but then that loads a bunch of assets async, including https://whoami.example.com/style.css, and my browser's request for style.css doesn't include the cookie (it should, but I do see safari is buggy here) - is it correct that you would also expect the viewport to redirect to authelia, even though the actual webpage loaded? What would authelia then redirect back to? The current page, not specifically style.css, I assume?

I am a little confused about the fetch API you are referring to - these are just included in raw html like:

```
<head>
<link rel="stylesheet" href="style.css">
</head>
```

So I'm not sure how fetchAPI comes into play here.

**Reply by james-d-elliott** [reply]
So if the browser loads them async without it being loaded via javascript it'll operate normally, or should per internet standards. The issue is when it's trying to load assets via javascript. If the javascript loading it uses the fetch API which is likely the case here, fetch not only will not properly follow the redirect, but actively prevents you getting the headers when the response code is 30x. I believe from my recent reading fetch also doesn't generally include cookies for these requests (so they will ALWAYS be unauthenticated requests, so unless you do bypass them they are possibly going to cause you lots of pain), so it may not even be Safari in this instance.

As far as if they are included statically in the HTML I have no clue, maybe some browsers are using fetch natively to do this? Can you replicate it in Chrome/Firefox? Is the original HTML document being cached?

Also for us to comment without wild guessing we're going to need a lot more information like logs and a [HAR](https://www.authelia.com/reference/guides/http-archive-file/) along with a screenshot showing the initiator.
